source/exe/signal_action.cc

```diff
@@ -42,21 +42,24 @@ void SignalAction::installSigHandlers() {
   stack.ss_size = altstack_size_;        // ... guard page at the other
   stack.ss_flags = 0;
 
-  RELEASE_ASSERT(sigaltstack(&stack, nullptr) == 0);
+  RELEASE_ASSERT(sigaltstack(&stack, &previous_altstack_) == 0);
 
+  int hidx = 0;
```

Comment:
For test purposes (and most slow path code in the production binary), I think in general minor performance optimization is second place to clarity, so the guiding principle is do what's simplest. 

I would have just used a `std::unordered_map<int, function pointer>`, but your code works as well, so ship it :)